app/src/main/java/org/oppia/android/app/hintsandsolution/HintsAndSolutionDialogFragmentPresenter.kt

```diff
@@ -328,46 +290,34 @@ class HintsAndSolutionDialogFragmentPresenter @Inject constructor(
 
     if (solutionViewModel.solutionCanBeRevealed.get()!!) {
       binding.root.visibility = View.VISIBLE
-      binding.revealSolutionButton.setOnClickListener {
+      binding.showSolutionButton.setOnClickListener {
         showRevealSolutionDialogFragment()
       }
     }
 
     binding.root.setOnClickListener {
       if (solutionViewModel.isSolutionRevealed.get()!!) {
-        val previousIndex: Int? = currentExpandedHintListIndex
-        currentExpandedHintListIndex =
-          if (currentExpandedHintListIndex != null && currentExpandedHintListIndex == position) {
-            null
-          } else {
-            position
-          }
-        expandedHintListIndexListener.onExpandListIconClicked(currentExpandedHintListIndex)
-        if (previousIndex != null &&
-          currentExpandedHintListIndex != null &&
-          previousIndex == currentExpandedHintListIndex
-        ) {
-          bindingAdapter.notifyItemChanged(currentExpandedHintListIndex!!)
-        } else {
-          previousIndex?.let {
-            bindingAdapter.notifyItemChanged(it)
-          }
-          currentExpandedHintListIndex?.let {
-            bindingAdapter.notifyItemChanged(it)
-          }
-        }
+        expandOrCollapseItem(position)
       }
     }
   }
 
+  private fun bindReturnToLessonViewModel(
+    binding: ReturnToLessonButtonItemBinding,
+    returnToLessonViewModel: ReturnToLessonViewModel
+  ) {
+    binding.buttonViewModel = returnToLessonViewModel
+
+    binding.returnToLessonButton.setOnClickListener {
+      (fragment.requireActivity() as? HintsAndSolutionListener)?.dismiss()
+    }
+  }
+
   private fun handleAllHintsExhausted(allHintsExhausted: Boolean) {
-    if (itemList[itemList.size - RECYCLERVIEW_INDEX_CORRECTION_MULTIPLIER] is SolutionViewModel) {
-      val solutionViewModel =
-        itemList[itemList.size - RECYCLERVIEW_INDEX_CORRECTION_MULTIPLIER] as SolutionViewModel
+    if (itemList[itemList.size - 2] is SolutionViewModel) {
```

Comment:
`2` because there is a button `Return to Lesson` at the bottom of the list, the index for which will be `itemList.size - 1` and therefore the index for solution will be `itemList.size - 2`

Added following comment above this code
```
    // The last item of the list is ReturnToLessonViewModel and therefore second last item is
    // SolutionViewModel as a result subtracting 2 from itemList size.
```